spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/core/subscriber/PubSubSubscriberOperations.java

```diff
@@ -114,27 +115,32 @@
 	PubsubMessage pullNext(String subscription);
 
 	/**
-	 * Acknowledge a batch of messages. The method will group the messages by subscription name
-	 * and acknowledge them in batches.
+	 * Acknowledge a batch of messages. The messages must have the same project id and be from the same subscription.
 	 * @param acknowledgeablePubsubMessages messages to be acknowledged
+	 * @return {@code ListenableFuture<Void>} the ListenableFuture for the asynchronous execution
+	 * @since 1.1
 	 */
-	void ack(Collection<AcknowledgeablePubsubMessage> acknowledgeablePubsubMessages);
+	ListenableFuture<Void> ack(Collection<AcknowledgeablePubsubMessage> acknowledgeablePubsubMessages);
 
 	/**
-	 * Negatively acknowledge a batch of messages. The method will group the messages by subscription name
-	 * and acknowledge them in batches.
+	 * Negatively acknowledge a batch of messages. The messages must have the same project id and be from the same
+	 * subscription.
 	 * @param acknowledgeablePubsubMessages messages to be negatively acknowledged
+	 * @return {@code ListenableFuture<Void>} the ListenableFuture for the asynchronous execution
+	 * @since 1.1
 	 */
-	void nack(Collection<AcknowledgeablePubsubMessage> acknowledgeablePubsubMessages);
+	ListenableFuture<Void> nack(Collection<AcknowledgeablePubsubMessage> acknowledgeablePubsubMessages);
 
 	/**
-	 * Modify the ack deadline of a batch of messages. The method will group the messages by subscription name
-	 * and modify their ack deadline in batches.
+	 * Modify the ack deadline of a batch of messages. The messages must have the same project id and be from the
```

Comment:
Why can't we just group automatically by `ProjectSubscriptionName` like before?

Reply:
Grouping by `ProjectSubscriptionName` doesn't solve the issue that messages need to have the same project id as the `PubSubSubscriber` as well as the same permissions to the subscription which is associated with the ackId.

Assuming the constraint above is met ...

Returning a `Collection` of `ListenableFuture`s is not enough information for developer to process / understand what happened / which messages were an issue. I don't see how that provides any value.

Completely my opinion, but I can see a developer thinking ...

"I made the method call with a single `Collection` of messages, but I get a `Collection` of `ListenableFuture`s as a return value.  I can see that the second `ListenableFuture` failed ... but what messages are in that list?"

Reply:
I wasn't thinking of returning a collection of `ListenableFuture`, but maybe rather `ListenableFuture<Collection<ProjectSubscriptionName>>` to indicate which groups of messages were acked.